### PR TITLE
Using a shared linux worker queue

### PR DESCRIPTION
**Pull request by 1Jo1**
Motivation:

on each eventloop 4 kthread worker will be created, to use the wq more efficiently you can attach different ringbuffers( (eventloops) to the same wq(worker queue)

Modifications:

- attach the wqFd to the ring buffer

Result:

using less worker queues

### CODE REVIEW

File: src/main/java/io/netty/incubator/channel/uring/Native.java [hunk 1]
```diff
@@ -35,6 +35,10 @@
     static final int DEFAULT_IOSEQ_ASYNC_THRESHOLD =
             Math.max(0, SystemPropertyUtil.getInt("io.netty.uring.iosqeAsyncThreshold", 25));
 
+    //the number of eventLoop which are associated to one work queue
```
Comment (martin-g) [inline src/main/java/io/netty/incubator/channel/uring/Native.java:38]: s/eventLoop/eventLoops/

**Comment by normanmaurer** [comment]
@1Jo1 you have any benchmarks to share here ?

File: src/main/c/netty_io_uring_native.c [hunk 2]
```diff
@@ -218,10 +218,13 @@ static jboolean netty_io_uring_probe(JNIEnv *env, jclass clazz, jint ring_fd, ji
...
 }
 
-
-static jobjectArray netty_io_uring_setup(JNIEnv *env, jclass clazz, jint entries) {
-    struct io_uring_params p;
-    memset(&p, 0, sizeof(p));
+static jobjectArray netty_io_uring_setup(JNIEnv *env, jclass clazz, jint entries, jint wqFd) {
+    struct io_uring_params params;
+    memset(&params, 0, sizeof(params));
+    if (wqFd > 0) {
```
Comment (normanmaurer) [inline src/main/c/netty_io_uring_native.c:224]: should we better use > -1 and use -1 if we don't want to use a wqFd. ?

Comment (normanmaurer) [reply src/main/c/netty_io_uring_native.c:224]: Actually it seems like we use -1 in the java code already. So I think just this if branch should be changed

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by 1Jo1** [comment]
yeah sure i’m gonna do it later

**Comment by 1Jo1** [comment]
@normanmaurer when benchmarking i noticed that the io workers are not destroyed(however ring fd are destroyed) this leads to the fact that i literally have to restart the vm to get rid of the worker queue because i have no memory...
let's see, I'll try other kernel version :)

**Comment by 1Jo1** [comment]
this PR is not related to this issue

**Comment by normanmaurer** [comment]
@1Jo1 I wonder if this is also the same as this one https://github.com/netty/netty-incubator-transport-io_uring/issues/14 ?

**Comment by 1Jo1** [comment]
@normanmaurer yeah probably, I'll take a look :)

**Comment by normanmaurer** [comment]
@1Jo1 did you find out anything yet ?

**Comment by 1Jo1** [comment]
@normanmaurer it's really weird...still trying to reproduce the bug in C...just need some time, a high number of requests and when `Native.eventFdWrite` is executed in a the wakeup method

**Comment by normanmaurer** [comment]
@1Jo1 now I guess we could try a benchmark again ?

**Comment by 1Jo1** [comment]
@normanmaurer yeah sure I'm gonna do that today :) sorry I'm kind of distracted as I'm still testing the fixed io_uring patches and I found another io_uring bug however it's not related to this PR & benchmark

**Comment by 1Jo1** [comment]
that's really strange, it seems to be that `lastRingFd` is zero in `newChild` method, but lastRingFd default is -1 @normanmaurer is reflection used to set all EventLoopGroup properties to zero?

**Comment by normanmaurer** [comment]
@1Jo1 no ... we don't use reflection

**Comment by 1Jo1** [comment]
@normanmaurer ok,  maybe lastRingFd is a race condition? :)

**Comment by normanmaurer** [comment]
`newChild(...)` should only be called by one thread.

**Comment by 1Jo1** [comment]
BTW I get this github actions error: `Error: servers must contain id, username and password` ? so I guess is not related this PR? as the tests run successfully on my machine

**Comment by normanmaurer** [comment]
@1Jo1 please rebase... I think this should fix it.

**Comment by 1Jo1** [comment]
@normanmaurer I already rebased https://github.com/1Jo1/netty-incubator-transport-io_uring/commits/shared-wq

**Comment by normanmaurer** [comment]
@1Jo1 ah wait... I did mess up the GitHub workflow. Gimme a few minutes

**Comment by 1Jo1** [comment]
no worries, take your time :)

**Comment by normanmaurer** [comment]
@1Jo1 please rebase again

**Comment by normanmaurer** [comment]
@1Jo1 anything to share yet ?

**Comment by 1Jo1** [comment]
my VM CPU has 4 cores

Benchmark cargo run --release -- --address "127.0.0.1:3000" --number 500 --duration 60 --length 500

DEFAULT_MAX_EVENTLOOPS_WQ: 5
the shared EventLoop uses 6% less memory because of less kernel threads

worker EventLoops: 3

500 clients, running 500 bytes, 60 sec.
Speed: 151461 request/sec, 151461 response/sec
Requests: 9087677
Responses: 9087675

worker EventLoops: 5

500 clients, running 500 bytes, 60 sec.
Speed: 166397 request/sec, 166397 response/sec
Requests: 9983852
Responses: 9983851


----------------------------------------------

without shared wq
DEFAULT_MAX_EVENTLOOPS_WQ: 1

worker EventLoops: 3
Speed: 181771 request/sec, 181770 response/sec
Requests: 10906262
Responses: 10906262


worker EventLoops: 5
cargo run --release -- --address "127.0.0.1:3000" --number 500 --duration 60 --length 500
Speed: 167598 request/sec, 167598 response/sec
Requests: 10055880
Responses: 10055880

**Comment by 1Jo1** [comment]
it depends how many CPU cores you have...I will do  benchmark tests tomorrow with a 6,8 or 12 Core VM

File: src/main/java/io/netty/incubator/channel/uring/Native.java [hunk 3]
```diff
@@ -35,6 +35,10 @@
     static final int DEFAULT_IOSEQ_ASYNC_THRESHOLD =
             Math.max(0, SystemPropertyUtil.getInt("io.netty.uring.iosqeAsyncThreshold", 25));
 
+    //the number of eventLoops which are associated to one work queue
```
Comment (normanmaurer) [inline src/main/java/io/netty/incubator/channel/uring/Native.java:38]: nit: `// The`

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by 1Jo1** [comment]
unfortunately I'm not able to connect to my workstation at home....I'm gonna run the benchmarks end of next week

**Comment by normanmaurer** [comment]
@1Jo1 any updates ?

**Comment by 1Jo1** [comment]
the benchmark are still unclear to me as more io-workers are created under load, 
I guess it's not dependent on the number of cores, maybe hard limit?
anyway digging into io-workers 

```
Benchmark cargo run --release -- --address "127.0.0.1:3000" --number 500 --duration 60 --length 500
# 6 Core VM

DEFAULT_MAX_EVENTLOOPS_WQ: 1

worker Eventloop: 3
500 clients, running 500 bytes, 60 sec.
Speed: 299178 request/sec, 299178 response/sec
Requests: 17950729
Responses: 17950727

worker EventLoops: 5
Speed: 370296 request/sec, 370296 response/sec
Requests: 22217796
Responses: 22217795

worker Eventloop: 6
Speed: 445790 request/sec, 445790 response/sec
Requests: 26747407
Responses: 26747404

------------------------------------------------------------------------------

DEFAULT_MAX_EVENTLOOPS_WQ: 5

worker EventLoops: 3
Speed: 279123 request/sec, 279123 response/sec
Requests: 16747398
Responses: 16747395

worker EventLoops: 5
Speed: 414602 request/sec, 414602 response/sec
Requests: 24876165
Responses: 24876160

worker EventLoops: 6
Speed: 505535 request/sec, 505535 response/sec
Requests: 30332156
Responses: 30332155

# 12 Core

DEFAULT_MAX_EVENTLOOPS_WQ: 1

EventLoops: 3
Speed: 411137 request/sec, 411137 response/sec
Requests: 24668253
Responses: 24668249

EventLoops: 5:
Speed: 518292 request/sec, 518292 response/sec
Requests: 31097556
Responses: 31097548

EventLoops: 6
Speed: 531681 request/sec, 531681 response/sec
Requests: 31900894
Responses: 31900886

DEFAULT_MAX_EVENTLOOPS_WQ: 5

EventLoops: 3
Speed: 391585 request/sec, 391585 response/sec
Requests: 23495105
Responses: 23495101

EventLoops: 5
Speed: 426210 request/sec, 426210 response/sec
Requests: 25572639
Responses: 25572637

EventLoops: 6
Speed: 426210 request/sec, 426210 response/sec
Requests: 25572639
Responses: 25572637
```

**Comment by 1Jo1** [comment]
@normanmaurer I did several performance tests...regarding memory usage it's really hard to measure how much memory io_uring is actually using, you can only measure the total kernel memory(as far as I know) I don't think there will a big benefit in terms of memory efficiency and performance, I came to the conclusion to close the PR for now
what do you think?

**Comment by normanmaurer** [comment]
let's close it for now I guess

**Pull request closed without merging**